agent/util-scripts/pbench-tool-meister-stop

```diff
@@ -66,86 +90,120 @@ def main(argv):
-    except Exception as e:
-        logger.error(
-            "Unable to connect to redis server, localhost:%d: %r", redis_port, e
-        )
-        return 2
-
-    try:
-        tm_pids_raw = redis_server.get("tm-pids")
-        if tm_pids_raw is None:
-            logger.error('Tool Meister PIDs key, "tm-pids", does not exist.')
-            return 3
-        tm_pids_str = tm_pids_raw.decode("utf-8")
-        tm_pids = json.loads(tm_pids_str)
-    except Exception as ex:
-        logger.error('Unable to fetch and decode "tm-pids" key: %s', ex)
-        return 4
     else:
-        expected_pids = 0
-        if "ds" in tm_pids:
-            expected_pids += 1
-        if "tm" in tm_pids:
-            expected_pids += len(tm_pids["tm"])
-        logger.debug("tm_pids = %r", tm_pids)
+        group = cli_params.tool_group
+
+    sysinfo, bad_l = cli_verify_sysinfo(cli_params.sysinfo)
+    if bad_l:
+        logger.error('invalid sysinfo option(s), "{}"', ",".join(bad_l))
 
-    ret_val = 0
+    interrupt = cli_params.interrupt
+    if interrupt and sysinfo:
+        # Don't collect system information if the CLI invocation specified an
+        # interrupt occured.
+        logger.warning("system information not collected when --interrupt specified")
+        sysinfo = ""
 
-    logger.debug("terminating %d pids at localhost:%d", expected_pids, redis_port)
-    terminate_msg = dict(action="terminate", group=None, directory=None, args=None)
     try:
-        num_present = redis_server.publish(
-            channel, json.dumps(terminate_msg, sort_keys=True)
-        )
+        full_hostname = os.environ["_pbench_full_hostname"]
+        benchmark_run_dir = Path(os.environ["benchmark_run_dir"]).resolve(strict=True)
     except Exception:
-        logger.exception("Failed to publish terminate message")
-        ret_val = 1
-    else:
-        if num_present != expected_pids:
+        logger.exception("failed to fetch parameters from the environment")
+        return 1
+
+    try:
+        redis_server_pid_file = (
+            benchmark_run_dir / "tm" / f"redis_{redis_port:d}.pid"
+        ).resolve(strict=True)
+    except FileNotFoundError:
+        # No Redis server, nothing to do.
+        return 0
+
+    # The Redis server is always running on the local host with the CLI.
+    with Client(
+        redis_host="localhost",
+        redis_port=redis_port,
+        channel_prefix=cli_tm_channel_prefix,
+        logger=logger,
+    ) as client:
+        # First we end the persistent tools
+        tool_dir = benchmark_run_dir / f"tools-{group}"
+        try:
+            tool_dir.mkdir(exist_ok=True)
+        except Exception as exc:
             logger.error(
-                "Failed to terminate %d pids, only encountered %d on the channel",
-                expected_pids,
-                num_present,
+                'failed to create tool output directory, "{}": {}', tool_dir, exc
```

Comment:
We can't just log an error and continue as if nothing happened, that is for sure.

Without the existence of the `tool_dir` directory, we can't do anything with the persistent tools, then `end` operation won't work because the Tool Data Sink won't have a directory to store any persistent data into.

Further, it is like that being unable to create the directory, the `sysinfo` operation will not work, so we really should just bail entirely here, returning an error, and requesting the entire tool meister infrastructure shutdown down, i.e. terminate.

Reply:
The `pbench-tool-meister-stop` command will now return an error when the `end` operation fails, or if we can't create the directory for the `end` operation to work.

We always wait for the local Tool Data Sink and local Tool Meister to exit before killing the Redis server, regardless of the success or failure of `terminate` operation.